### ISSUE TITLE
feat: add more auth options to the CLI

### ISSUE THREAD

**Issue by MahmoudESSE** [t3-oss/create-t3-app#1593]
### Is your feature request related to a problem? Please describe.

I noticed that we only use [nextAuth](https://next-auth.js.org/) or choose none to start with no auth, then either add an auth service manually or develop without auth.

### Describe the solution you'd like to see

Add [clerk](https://clerk.com/) to auth option as a start just like drizzle was added to the ORM list to quickly setup clerk using the CLI instead of manual setup.

### Describe alternate solutions

.

### Additional information

Maybe later add other auth method besides clerk and nextAuth but for now this one the ones I see been used often, so I wanted to know if this is a desirable feature to add to the CLI.

**Comment by mqshaikh8** [comment]
This sounds interesting. I wonder if that is in the plan for this tool. I would love to do this if it something that is planned and no takers.

**Comment by juliusmarminge** [comment]
We are considering adding Clerk and/or Lucia but still nothing definite.

**Comment by ronanru** [comment]
I really like Lucia for auth, can I make a PR that adds it as an option?

**Comment by MahmoudESSE** [comment]
I read on an old issue that there were plans for "add on" plugin for auth, can you provide examples or idea of how we should implement it

**Comment by juliusmarminge** [comment]
cc @c-ehrlich @nexxeln what'ya think?

**Comment by dBianchii** [comment]
I thought that one big ethos of t3-oss was to honor the last part of the name... -oss.

Although, I see that if we still have the option to choose open-source alternatives, then technically it is still honoring that

**Comment by c-ehrlich** [comment]
> > I really like Lucia for auth, can I make a PR that adds it as an option?
> 
> cc @c-ehrlich @nexxeln what'ya think?

id be up for it. but @ronanru maybe first make an example repo(s) for what output apps would look like - to keep it small, maybe /pages + prisma and /app + drizzle. gives us something to look at / talk about with less initial effort.

**Comment by ronanru** [comment]
> > > I really like Lucia for auth, can I make a PR that adds it as an option?
> > 
> > 
> > cc @c-ehrlich @nexxeln what'ya think?
> 
> id be up for it. but @ronanru maybe first make an example repo(s) for what output apps would look like - to keep it small, maybe /pages + prisma and /app + drizzle. gives us something to look at / talk about with less initial effort.

Made a quick output example. 
/app + drizzle and /pages + prisma + trpc
https://github.com/ronanru/t3-lucia

**Comment by raphaelm-gioa** [comment]
Waiting on Auth.js to drop so we have a Expo/Next.js connector merged to it - and finally have a React Native/React connector !
Lucia sounds fair also ! I believe ct3a should support Clerk as well as self-hosted solutions - we're all about no lock-in ! (Using Clerk in all my pro projects btw, no shade here, it's an amazing tool)

**Comment by RobSchilderr** [comment]
What about Supertokens? It's similar to Clerk but has a self-hosted option.

**Comment by rishabhpoddar** [comment]
> What about Supertokens? It's similar to Clerk but has a self-hosted option.

As a follow on to what @RobSchilderr said, the template integration can use `https://try.supertokens.com` as the SuperTokens core service, so that auth works out of the box when someone generates a new app. Then users can switch to their instance of the core (managed or self hosted) via a config change.

**Comment by Wundero** [comment]
> As a follow on to what @RobSchilderr said, we can also offer users the choice of adding different types of auth to their app:
> 
> * email password
> * social login
>   
>   * For social, we can also offer them a pick of google, fb, github oauth and provide dev credentials which work right out of the box (which users should replace before pushing to prod)
> * passwordless (magic links, OTP)
> * any combination of the above
> 
> This choice can be from the CLI itself, and the generated app has the relevant config already setup for their chosen auth method (and in case users are not sure which method they want, we can just default to email password).
> 
> Regardless of SuperTokens being added, I think allowing users to pick a specific auth method will be cool.

I personally don't see why this would be a necessary addition to the CLI tool - NextAuth, Lucia, etc. all provide fairly straightforward ways to customize the providers (email, oauth, etc.) that you can use as they are, so having a choice in the CLI tool saves very little time and bloats the code of create-t3-app. Additionally, you can usually use more than one at once, and that increases the complexity further of integrating that into the CLI. IMO good examples and documentation are better in this case than more CLI options.

Separately, providing "dev credentials" for oauth isn't really feasible, since oauth requires you to setup your app by the providers, and even if it was feasible, it would be pretty insecure. oauth is really strong because it's really easy to use to sign in with real accounts, so dev credentials can just be whatever provider you use (github, discord, etc.).

**Comment by rishabhpoddar** [comment]
> NextAuth, Lucia, etc. all provide fairly straightforward ways to customize the providers (email, oauth, etc.)

Yup, however, it's just one less friction point for users to get what they want. In order to customise these solutions, users would need to dive into the docs of these auth providers, which can be annoying. But I understand that it will make the CLI code a bit more complex.

> Separately, providing "dev credentials" for oauth isn't really feasible, since oauth requires you to setup your app by the providers, and even if it was feasible, it would be pretty insecure

There is a way to make it work for any domain by introducing a proxy domain in the middle. This IS a security issue indeed, which is why we should somehow enforce that these credentials are only used for dev builds. Again, the aim here is to get people started as quickly as possible.

**Comment by c-ehrlich** [comment]
we are definitely never adding credentials auth. if someone wants to implement that, that's on them. the effort of reading one page in the next-auth docs is tiny compared to the actual work that is required to make something that is secure and fits the greater needs of their project. having it as a CLI options is leading people down a dangerous path.

**Comment by juliusmarminge** [comment]
> What about Supertokens? It's similar to Clerk but has a self-hosted option.

I dont think anyone on the core team has tried that and we're only adding stuff we're comfortable to recommend to others so it's currently Clerk and NextAuth, maybe Lucia

**Comment by henrikvtcodes** [comment]
I would like to +1 Lucia here, since it seems to fit well within T3, specifically because it is similar to NextAuth/Auth.js in that it runs within Next.js - a notable difference between it and something like Clerk, Supertokens, etc. Adding Clerk or a different hosted auth provider seems like it would overcomplicate things, because the end-user/developer would have to go set it up and get API keys for the hosted service in addition to getting the Client ID and secret for their OAuth provider(s) of choice before they can run the dev script and get going.

**Comment by rishabhpoddar** [comment]
> in addition to getting the Client ID and secret for their OAuth provider(s) of choice before they can run the dev script and get going.

Won't that be true for any auth solution (including nextauth)? Unless of course, the generated app uses oauth credentials that relies on a proxy domain.

**Comment by henrikvtcodes** [comment]
> Won't that be true for any auth solution (including nextauth)? Unless of course, the generated app uses oauth credentials that relies on a proxy domain.

Well yeah you always need a client ID and secret no matter what oauth solution you use, but my point was that NextAuth/Lucia dont need any other API keys or anything since they are not external to Next.js.

**Comment by cfarvidson** [comment]
I would like a clerk add on. (I have never tried Lucia)

I just finished adding clerk to a T3 project (trpc, typescript, prisma, tailwind) and it was not trivial (for me at least) to get everything to work. And by work I mean working auth in both client and server side components when using trpc queries.

**Comment by juliusmarminge** [comment]
Hey everyone,

We have decided not to pursue adding Lucia auth as we believe it's too similar to next-auth (especially with v5 around the corner and it's auth.js port).

Will keep this issue open until we've made a definitive answer on adding clerk or not.

**Comment by RobertHH-IS** [comment]
@cfarvidson I am also having a difficult time integrating clerk into the unique t3 trpc setup.  Any way you would post the main parts of the deployment?

**Comment by rishabhpoddar** [comment]
Is there interest in adding other auth providers like [SuperTokens](https://supertokens.com)?

**Comment by ielm** [comment]
@RobertHH-IS there's an example project by @perkinsjr for integrating clerk into t3 trpc [here](https://github.com/perkinsjr/t3-app-clerk-minimal/tree/main). It's based on the pages router, but still helpful. 

I would also like to +1 adding clerk auth to create-t3-app. I think most people using t3 are opting for next-auth or clerk, and getting clerk to play nicely with trpc is a little painful. 

For anyone interested in this in the meantime, clerk provides documentation for pages router integration [here](https://clerk.com/docs/references/nextjs/trpc).

**Comment by rpatel15-hue** [comment]
Here are my thoughts: Many users of create-t3 are focused on quickly shipping and iterating on their projects. A challenge emerges, for example, when a project scales and there's a shift towards developing a React Native application. At this point, next-auth proves to be incompatible with React Native. Introducing support for alternatives like Lucia, and possibly Clerk (despite it not being open-source), could offer a seamless transition. This strategy would prevent users from having to undertake extensive refactoring when they decide to expand their projects.

**Comment by juliusmarminge** [comment]
> A challenge emerges, for example, when a project scales and there's a shift towards developing a React Native application. At this point, next-auth proves to be incompatible with React Native. Introducing support for alternatives like Lucia

This statement is just not true. next-auth supports react native auth in the exact same way as lucia does

**Comment by rpatel15-hue** [comment]
@juliusmarminge You are right, I knew about `next-auth` but did not know about `lucia`. For some reason, I thought they did. Here are the two sources I read:

- Next Auth: [https://github.com/nextauthjs/next-auth/blob/72ba83c3ee9a78b1bccac2a2629b76a073ede3be/docs/docs/concepts/faq.md?plain=1#L78](https://github.com/nextauthjs/next-auth/blob/72ba83c3ee9a78b1bccac2a2629b76a073ede3be/docs/docs/concepts/faq.md?plain=1#L78)
- Lucia: [https://github.com/lucia-auth/lucia/discussions/652](https://github.com/lucia-auth/lucia/discussions/652)

**Comment by Syarx** [comment]
I am willing to try and implement lucia in ct3a as it is at the moment, would that be allowed to merge if it is up to the standards?

**Comment by ronanru** [comment]
> I am willing to try and implement lucia in ct3a as it is at the moment, would that be allowed to merge if it is up to the standards?

see https://github.com/t3-oss/create-t3-app/issues/1593?notification_referrer_id=NT_kwDOAh-9CrM4MDU1MTcxMDk4OjM1NjM0NDQy#issuecomment-1817808074

**Comment by ronanru** [comment]
> I am willing to try and implement lucia in ct3a as it is at the moment, would that be allowed to merge if it is up to the standards?

https://github.com/t3-oss/create-t3-app/issues/1593#issuecomment-1817808074

**Comment by lorensr** [comment]
@juliusmarminge would a PR adding a Clerk option be welcome?

**Comment by astridlol** [comment]
> @juliusmarminge would a PR adding a Clerk option be welcome?

Clerk is so incredibly easy to set up, that I don't really see a reason for this.
It's as simple as pasting two env vars, adding a middleware.ts and putting ClerkProvider in your root layout or _app.ts.

**Comment by Chenalejandro** [comment]
Hey @juliusmarminge! I just watched the last Theo video talking about this. Instead of Clerk, I would like to propose a new managed auth: [stack-auth](https://github.com/stack-auth/stack). It is as easy to set up as Clerk, and... it is fully open sourced. You can even host it yourself.

I just migrated my project from next-auth to stack-auth, and it has been great! Now I can support email/password authentication, and have a separate sign-in/sign-up page!

Clerk is obviously ahead in terms of functionality and/or documentation, but in my opinion the clouse-source nature of it doesn't fit the t3 stack well. And stack-auth is advancing quickly and we can even contribute it to make it better still!

BTW, it is also backed by the creator of next.js, Guillermo Rauch

**Comment by lacymorrow** [comment]
You can have separate sign in/up pages with auth js/next-auth. Not saying it's good, but it isn't a "missing feature"

**Comment by moebqr** [comment]
Really really appreciate if we can get Clerk as an auth option in Create-T3-App!!!! Some of the options mentioned here are not reasonable, but I think integration Clerk (and possibly Lucia) is a good option that will save devs hours.

**Comment by FleetAdmiralJakob** [comment]
Lucia will not be supported any longer: https://github.com/lucia-auth/lucia/discussions/1714

**Comment by GustavoOS** [comment]
Clerk is a Theo recommendation for over an year, this should really be there, as it is kind of complicated integrating it into trpc
Auth0 would be a good addition as well
